arangod/Cluster/ClusterMethods.cpp

```diff
@@ -753,251 +839,262 @@ bool smartJoinAttributeChanged(LogicalCollection const& collection, VPackSlice c
-                              // the DBserver could have reported an error.
+    };
+    auto pre = [](Result&, VPackBuilder& builder) -> void {
+      // initialize to empty object
+      builder.openObject();
+      builder.close();
+    };
+    return handleResponsesFromAllShards(results, handler, pre);
+  };
+  return futures::collectAll(std::move(futures)).thenValue(std::move(cb));
 }
 
 ////////////////////////////////////////////////////////////////////////////////
 /// @brief counts number of documents in a coordinator, by shard
 ////////////////////////////////////////////////////////////////////////////////
 
-int countOnCoordinator(transaction::Methods& trx, std::string const& cname,
-                       std::vector<std::pair<std::string, uint64_t>>& result) {
+futures::Future<OperationResult> countOnCoordinator(transaction::Methods& trx,
+                                                    std::string const& cname) {
+  std::vector<std::pair<std::string, uint64_t>> result;
+
   // Set a few variables needed for our work:
   ClusterInfo* ci = ClusterInfo::instance();
   auto cc = ClusterComm::instance();
   if (cc == nullptr) {
     // nullptr happens only during controlled shutdown
-    return TRI_ERROR_SHUTTING_DOWN;
+    return futures::makeFuture(OperationResult(TRI_ERROR_SHUTTING_DOWN));
   }
 
-  result.clear();
-
   std::string const& dbname = trx.vocbase().name();
   // First determine the collection ID from the name:
   std::shared_ptr<LogicalCollection> collinfo;
   collinfo = ci->getCollectionNT(dbname, cname);
   if (collinfo == nullptr) {
-    return TRI_ERROR_ARANGO_DATA_SOURCE_NOT_FOUND;
+    return futures::makeFuture(OperationResult(TRI_ERROR_ARANGO_DATA_SOURCE_NOT_FOUND));
   }
 
   std::shared_ptr<ShardMap> shardIds = collinfo->shardIds();
   const bool isManaged = trx.state()->hasHint(transaction::Hints::Hint::GLOBAL_MANAGED);
   if (isManaged) {
     Result res = ::beginTransactionOnAllLeaders(trx, *shardIds);
     if (res.fail()) {
-      return res.errorNumber();
+      return futures::makeFuture(OperationResult(res));
     }
   }
 
-  std::vector<ClusterCommRequest> requests;
-  auto body = std::make_shared<std::string>();
+  std::vector<Future<network::Response>> futures;
+  futures.reserve(shardIds->size());
+
   for (std::pair<ShardID, std::vector<ServerID>> const& p : *shardIds) {
-    auto headers = std::make_unique<std::unordered_map<std::string, std::string>>();
-    ClusterTrxMethods::addTransactionHeader(trx, /*leader*/ p.second[0], *headers);
-    requests.emplace_back("shard:" + p.first, arangodb::rest::RequestType::GET,
-                          "/_db/" + StringUtils::urlEncode(dbname) +
-                              "/_api/collection/" +
-                              StringUtils::urlEncode(p.first) + "/count",
-                          body, std::move(headers));
+    network::Headers headers;
+    addTransactionHeaderForShard(trx, *shardIds, /*shard*/ p.first, headers);
```

Comment:
why did you change it addTransactionHeaderForShard ?
```suggestion
    ClusterTrxMethods::addTransactionHeader(trx, /*leader*/ p.second[0], *headers);
```

Reply:
Sloppy copy/paste/edit. Will revert.